File: runtime/src/main/java/org/apache/flink/agents/runtime/client/AgentPlanClient.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.agents.runtime.client;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.flink.agents.plan.AgentPlan;
import org.apache.flink.agents.runtime.CompileUtils;
import org.apache.flink.agents.runtime.operator.coordinator.PlanUpdateMessages.PlanUpdateRequest;
import org.apache.flink.agents.runtime.operator.coordinator.PlanUpdateMessages.PlanUpdateResponse;
import org.apache.flink.annotation.VisibleForTesting;
import org.apache.flink.api.common.JobID;
import org.apache.flink.client.program.rest.RestClusterClient;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.configuration.RestOptions;
import org.apache.flink.runtime.operators.coordination.CoordinationResponse;
import org.apache.flink.util.Preconditions;

import java.util.concurrent.ExecutionException;

/** Client for submitting a complete AgentPlan update to a running Flink Agents job. */
public final class AgentPlanClient implements AutoCloseable {

    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
    private static final String CLUSTER_ID = "flink-agents-plan-client";

    private final RestClusterClient<String> restClient;

    /** Creates a client for a JobManager REST host and port. */
    public AgentPlanClient(String restAddress, int restPort) throws Exception {
        this(new AgentRestClusterClient(createRestConfiguration(restAddress, restPort)));
    }

    /** Reflection entry point used by the Python client. */
    private static AgentPlanClient create(String restAddress, Integer restPort) throws Exception {
        Preconditions.checkNotNull(restPort, "REST port must not be null");
        return new AgentPlanClient(restAddress, restPort);
    }

    @VisibleForTesting
    AgentPlanClient(RestClusterClient<String> restClient) {
        this.restClient = Preconditions.checkNotNull(restClient, "restClient must not be null");
    }

    private static Configuration createRestConfiguration(String restAddress, int restPort) {
        Preconditions.checkArgument(
                restAddress != null && !restAddress.trim().isEmpty(),
                "REST address must not be blank");
        Preconditions.checkArgument(
                restPort > 0 && restPort <= 65535,
                "REST port must be between 1 and 65535: %s",
                restPort);
        Configuration configuration = new Configuration();
        configuration.set(RestOptions.ADDRESS, restAddress.trim());
        configuration.set(RestOptions.PORT, restPort);
        return configuration;
    }

    /** Serializes and submits a complete plan for the named deployed agent. */
    public UpdateResult updateAgentPlan(JobID jobId, String agentName, AgentPlan agentPlan)
            throws Exception {
        Preconditions.checkNotNull(agentPlan, "agentPlan must not be null");
        return updateAgentPlan(jobId, agentName, OBJECT_MAPPER.writeValueAsString(agentPlan));
    }

    /** Submits a complete plan JSON document for the named deployed agent. */
    public UpdateResult updateAgentPlan(JobID jobId, String agentName, String agentPlanJson)
            throws Exception {
        Preconditions.checkNotNull(jobId, "jobId must not be null");
        Preconditions.checkNotNull(agentPlanJson, "agentPlanJson must not be null");
        String operatorUid = CompileUtils.getCoordinatedOperatorUid(agentName);
        CoordinationResponse response;
        try {
            response =
                    CoordinationRequestUtils.send(
                                    restClient,
                                    jobId,
                                    operatorUid,
                                    new PlanUpdateRequest(agentPlanJson))
                            .get();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw e;
        } catch (ExecutionException e) {
            throw unwrapExecutionException(e);
        }
        if (!(response instanceof PlanUpdateResponse)) {
            throw new IllegalStateException(
                    "Expected PlanUpdateResponse but received "
                            + (response == null ? "null" : response.getClass().getName()));
        }
        PlanUpdateResponse planUpdateResponse = (PlanUpdateResponse) response;
        return new UpdateResult(
                planUpdateResponse.planVersion,
                planUpdateResponse.planId,
                planUpdateResponse.subtasksNotified);
    }

    /** String-only bridge used by Python's reflection helper. */
    public UpdateResult updateAgentPlan(String jobId, String agentName, String agentPlanJson)
            throws Exception {
        Preconditions.checkNotNull(jobId, "jobId must not be null");
        return updateAgentPlan(JobID.fromHexString(jobId), agentName, agentPlanJson);
    }

    private static Exception unwrapExecutionException(ExecutionException exception) {
        Throwable cause = exception.getCause();
        if (cause instanceof Exception) {
            return (Exception) cause;
        }
        if (cause instanceof Error) {
            throw (Error) cause;
        }
        return new RuntimeException(cause);
    }

    @Override
    public void close() throws Exception {
        restClient.close();
    }

    /**
     * Confirmation that the current coordinator attempt accepted an update as a volatile candidate.
     * It does not mean the update was announced, made durable, or activated.
     */
    public static final class UpdateResult {
        private final long planVersion;
        private final String planId;
        private final int registeredSubtasks;

        private UpdateResult(long planVersion, String planId, int registeredSubtasks) {
            this.planVersion = planVersion;
            this.planId = planId;
            this.registeredSubtasks = registeredSubtasks;
        }

        public long getPlanVersion() {
            return planVersion;
        }

        public String getPlanId() {
            return planId;
        }

        /** Number of subtask gateways registered when the update was accepted. */
        public int getRegisteredSubtasks() {
            return registeredSubtasks;
        }
    }

    /** Ensures coordination responses are deserialized with the Flink Agents classloader. */
    private static final class AgentRestClusterClient extends RestClusterClient<String> {
        private AgentRestClusterClient(Configuration configuration) throws Exception {
            super(configuration, CLUSTER_ID);
        }
    }
}

File: python/flink_agents/api/agent_plan_client.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
import importlib
from abc import ABC, abstractmethod
from dataclasses import dataclass
from types import TracebackType
from typing import TYPE_CHECKING, Any, Mapping, Type

from typing_extensions import Self

from flink_agents.api._java_utils import (
    add_flink_agents_jars_to_context_class_loader,
)

if TYPE_CHECKING:
    from flink_agents.api.agents.agent import Agent


@dataclass(frozen=True)
class PlanUpdateResult:
    """Confirmation that the current JobManager attempt accepted an AgentPlan.

    ``registered_subtasks`` is the number of subtask gateways registered when
    the request was accepted as a volatile candidate. It does not mean the plan
    was announced, made durable, or became effective.
    """

    plan_version: int
    plan_id: str
    registered_subtasks: int


class AgentPlanClient(ABC):
    """Client for submitting AgentPlan updates to a running Flink job."""

    @staticmethod
    def create(rest_address: str, rest_port: int) -> "AgentPlanClient":
        """Create a client for a Flink JobManager REST endpoint.

        Args:
            rest_address: JobManager hostname or IP address, without a URL scheme
                or path.
            rest_port: JobManager REST port in the range 1 through 65535.

        Raises:
            ValueError: If the address is blank or the port is outside the valid
                range.
        """
        if not rest_address or not rest_address.strip():
            msg = "The REST address must not be blank."
            raise ValueError(msg)
        if rest_port <= 0 or rest_port > 65535:
            msg = "The REST port must be between 1 and 65535."
            raise ValueError(msg)
        add_flink_agents_jars_to_context_class_loader()
        runtime_module = importlib.import_module(
            "flink_agents.runtime.agent_plan_client"
        )
        return runtime_module.create_instance(rest_address, rest_port)

    @abstractmethod
    def update_agent_plan(
        self, job_id: str, agent_name: str, agent_plan_json: str
    ) -> PlanUpdateResult:
        """Validate and submit a complete AgentPlan JSON document."""

    @abstractmethod
    def update_agent(
        self,
        job_id: str,
        agent: "Agent",
        name: str | None = None,
        config: Mapping[str, Any] | None = None,
    ) -> PlanUpdateResult:
        """Compile an Agent with optional config data and submit it."""

    @abstractmethod
    def close(self) -> None:
        """Close the underlying REST client."""

    def __enter__(self) -> Self:
        """Return this client for use as a context manager."""
        return self

    def __exit__(
        self,
        exc_type: Type[BaseException] | None,
        exc_value: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        """Close this client when leaving a context manager."""
        self.close()
